### ISSUE TITLE
AWS CLI v2 installation directory permissions are broken in 2.1.4

### ISSUE THREAD

**Issue by bgray-onica** [aws/aws-cli#5769]
Confirm by changing [ ] to [x] below to ensure that it's a bug:
- [x] I've gone though the [User Guide](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-welcome.html) and the [API reference](https://docs.aws.amazon.com/cli/latest/reference/)
- [x] I've searched for [previous similar issues](https://github.com/aws/aws-cli/issues) and didn't find any solution
  
**Describe the bug**
Under a new install of 2.1.4, the folder permissions are wrong. This bug causes the PATH var to technically not work unless su into root.

This seems to be recent as I've not experienced this when installing a few 

**SDK version number**

**Platform/OS/Hardware/Device**
CentOS 8

**To Reproduce (observed behavior)**
1. Follow the instructions to [install](https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2-linux.html)
2. Run `aws`
3. Get command not found

**Expected behavior**
1. Follow instructions to [install](https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2-linux.html)
2. Run `aws`
3. Get the help output

**Logs/output**
Post-install command test
```
$ sudo ./aws/install
You can now run: /usr/local/bin/aws --version
$ aws
-bash: aws: command not found
```

Post install directory permissions
```
$ ll /usr/local/
drwx------. 3 root root 16 Nov 30 20:03 aws-cli
```

Post-install directory permissions FIXED
```
$ sudo chmod -R 755 aws-cli
$ aws

usage: aws [options] <command> <subcommand> [<subcommand> ...] [parameters]
To see help text, you can run:

  aws help
  aws <command> help
  aws <command> <subcommand> help

aws: error: the following arguments are required: command
```

**Additional context**
This was on a clean brand new instance of CentOS 8 using the CIS benchmark level 1 image.

**Comment by kdaily** [comment]
Hi @bgray-onica,

I tried to reproduce this on an Amazon Linux EC2, without success:

```
[ec2-user@xxxxxx ~]$ aws --version
aws-cli/2.1.4 Python/3.7.3 Linux/4.14.203-156.332.amzn2.x86_64 exe/x86_64.amzn.2
[ec2-user@xxxxxx ~]$ ll /usr/local/bin/
total 0
lrwxrwxrwx 1 root root 37 Nov 30 20:48 aws -> /usr/local/aws-cli/v2/current/bin/aws
lrwxrwxrwx 1 root root 47 Nov 30 20:48 aws_completer -> /usr/local/aws-cli/v2/current/bin/aws_completer
```

I'll look into it some more on a different OS.

**Comment by bgray-onica** [comment]
The symlinks will always show those permissions. That's just a nuance about symlinks. Look at the linked folder. Here's some examples of symlinks in the `/usr/bin` folder

```
lrwxrwxrwx. 1 root root      14 May 14  2019  jsondiff -> ./jsondiff-3.6
lrwxrwxrwx. 1 root root      14 May 14  2019  jsondiff-3 -> ./jsondiff-3.6
-rwxr-xr-x. 1 root root    1033 May 14  2019  jsondiff-3.6
lrwxrwxrwx. 1 root root      15 May 14  2019  jsonpatch -> ./jsonpatch-3.6
lrwxrwxrwx. 1 root root      15 May 14  2019  jsonpatch-3 -> ./jsonpatch-3.6
-rwxr-xr-x. 1 root root    3676 May 14  2019  jsonpatch-3.6
lrwxrwxrwx. 1 root root      17 May 14  2019  jsonpointer -> ./jsonpointer-3.6
lrwxrwxrwx. 1 root root      17 May 14  2019  jsonpointer-3 -> ./jsonpointer-3.6
```

And here's the same symlinks on the same instance I tested on (which are identical to yours and I did not modify permissions).
```
$ ll
lrwxrwxrwx. 1 root root 37 Nov 30 19:46 aws -> /usr/local/aws-cli/v2/current/bin/aws
lrwxrwxrwx. 1 root root 47 Nov 30 19:46 aws_completer -> /usr/local/aws-cli/v2/current/bin/aws_completer
```

**Comment by kdaily** [comment]
Hi @bgray-onica, did this issue not occur with a previous version of AWS CLI v2? I am aware of other issues related to hardened OSes. For example, the inability to access shared memory (`/dev/shm`) can impede some Python libraries from functioning properly.

**Comment by bgray-onica** [comment]
I have never experienced this issue before with v2 and it was puzzling me at first. It's always been available immediately without need to change dir perms.

**Comment by bgray-onica** [comment]
@kdaily Well I did some digging and found the issue but this should be handled in the install script to avoid the issue permanently. I would likely detect the umask early and gate a chmod based on the result, but there's tons of ways to execute the same result.

```
$ grep UMASK /etc/login.defs
UMASK           077
```

And here's the documentation of what 077 means:
https://www.cyberciti.biz/tips/understanding-linux-unix-umask-value-usage.html

So umask is giving zero permissions to all but owner by default. Should you find this should be closed and/or spawn a new issue, do whatever is necessary. Sorry I took up some time.

**Comment by kdaily** [comment]
Hi @bgray-onica, I appreciate you bringing this issue to our attention. It looks like based on the article you shared that the expected default is `022` for root, which I confirmed on RedHat 8. I imagine setting it to `077` is a hardening procedure, so in this case the installing user would be responsible for overriding that value to be sure it's desired, and that behavior on your system is expected.

**Comment by bgray-onica** [comment]
Yep unfortunately this is in a govcloud use case and hardened images are generally a requirement.

**Comment by kdaily** [comment]
Hi @bgray-onica, we're going to look into what can be improved. A section in the installation or troubleshooting guide can be added, and possibly a warning printed if after installation the permissions would not allow execution. One issue with the v2 is that there are more executables than the `aws` one.

**Comment by bgray-onica** [comment]
Any of those options I think are good ones too. I usually favor options that remove the need for the end user to even _know_ about the issue. Especially since the instructions already direct one to use `sudo` in the installation. Though I can understand the hesitation and desire to be transparent about it.

**Comment by roccomuscaritolo-okta** [comment]
I can confirm, I am running into the same issue when using `sudo ./aws/install`

The `aws` command only works from sudo or root.
```
roccomuscaritolo@G6BN93800A64:~ $ sudo aws --version
[sudo] password for roccomuscaritolo: 
aws-cli/2.2.15 Python/3.8.8 Linux/5.4.0-74-generic exe/x86_64.ubuntu.20 prompt/off
```

This is because of two separate issues.

1. The `aws-cli` directory is missing read and execute for the `others` permission group. (same issue with each sub-folder within it) preventing directory traversal.
```
drwxr-x--- 3 root root 4.0K Jul  1 15:23 /usr/local/aws-cli/
```

2. The aws binary itself (and every other file here) are also missing read and execute for the `others` permission group.
```
ls -lh /usr/local/aws-cli/v2/current/dist/aws
-rwxr-x--- 1 root root 4.3M Jul  1 15:23 /usr/local/aws-cli/v2/current/dist/aws
```

To fix this, you can add the missing perms recursively with `chmod -R`.
```
chmod -R o+rx /usr/local/aws-cli/
```

And now the command will work as intended.
```
roccomuscaritolo@G6BN93800A64:~ $ aws --version
aws-cli/2.2.15 Python/3.8.8 Linux/5.4.0-74-generic exe/x86_64.ubuntu.20 prompt/off
```

**Comment by vcsekhar** [comment]
Just wanted to share an update - 

Mac OS Big Sur V 11.6 - M1 processor > Reported issue still persists and above suggestions (
1. sudo chmod -R 755 aws-cli (from post dated 03 Nov 2020)
2. chmod -R o+rx /usr/local/aws-cli/ (from post dated 1 Jul 2021)
) didn't work. 

My Mac still keeps giving below output for aws --version

zsh: command not found: aws

**Comment by bgray-onica** [comment]
Sounds like it's a different issue then and you should open a new issue instead of piggybacking on this one @vcsekhar. This specifically deals with Linux and umask differences.

**Comment by cipri-tom** [comment]
This is really problematic beyond the 3-4 users who commented here. This question on ServerFault has 5000 views as of today: https://serverfault.com/questions/1034762/awscli-v2-needs-sudo-to-run-otherwise-get-permission-denied-error-ubuntu-20-04-0 

So it affects ubuntu too, not just CentOS

**Comment by bgray-onica** [comment]
@cipri-tom  That is not the same issue. Sorry but that needs a new issue. `Permission denied` is not the same as `Command not found`.

**Comment by sosafe-bastian-bretagne** [comment]
faced the same issue trying to install the cli in the python ms vscode dev container image, umask 022, I have set +x perm to the local/bin/aws file to solve my issue, not installing it with sudo tho.
Too bad there is no binary for the cli2 - still, its such a pain to install and to trust a 3rd party bash install script
